modules/objdetect/src/qrcode.cpp

```diff
@@ -949,34 +949,99 @@ vector<Point2f> QRDetect::getQuadrilateral(vector<Point2f> angle_list)
+
+bool QRCodeDetectorBase::detect(InputArray img, OutputArray points) const {
+    CV_Assert(p != nullptr);
+    return p->detect(img, points);
+}
+
+std::string QRCodeDetectorBase::decode(InputArray img, InputArray points, OutputArray straight_qrcode) {
+    CV_Assert(p != nullptr);
+    return p->decode(img, points, straight_qrcode);
+}
 
-struct QRCodeDetector::Impl
+std::string QRCodeDetectorBase::detectAndDecode(InputArray img, OutputArray points, OutputArray straight_qrcode) {
+    CV_Assert(p != nullptr);
+    return p->detectAndDecode(img, points, straight_qrcode);
+}
+
+bool QRCodeDetectorBase::detectMulti(InputArray img, OutputArray points) const {
+    CV_Assert(p != nullptr);
+    return p->detectMulti(img, points);
+}
+
+bool QRCodeDetectorBase::decodeMulti(InputArray img, InputArray points, std::vector<std::string>& decoded_info,
+                                     OutputArrayOfArrays straight_qrcode) const {
+    CV_Assert(p != nullptr);
+    return p->decodeMulti(img, points, decoded_info, straight_qrcode);
+}
+
+bool QRCodeDetectorBase::detectAndDecodeMulti(InputArray img, std::vector<std::string>& decoded_info,
+                                              OutputArray points, OutputArrayOfArrays straight_qrcode) const {
+    CV_Assert(p != nullptr);
+    return p->detectAndDecodeMulti(img, decoded_info, points, straight_qrcode);
+}
+
+struct ImplContour : public QRCodeDetectorBase::Impl
 {
 public:
-    Impl() { epsX = 0.2; epsY = 0.1; }
-    ~Impl() {}
+    ImplContour(): epsX(0.2), epsY(0.1)  {}
 
     double epsX, epsY;
-    vector<vector<Point2f>> alignmentMarkers;
-    vector<Point2f> updateQrCorners;
+    mutable vector<vector<Point2f>> alignmentMarkers;
+    mutable vector<Point2f> updateQrCorners;
     bool useAlignmentMarkers = true;
+
+    bool detect(InputArray in, OutputArray points) const override;
+    std::string decode(InputArray img, InputArray points, OutputArray straight_qrcode) override;
+    std::string detectAndDecode(InputArray img, OutputArray points, OutputArray straight_qrcode) override;
+
+    bool detectMulti(InputArray img, OutputArray points) const override;
+    bool decodeMulti(InputArray img, InputArray points, std::vector<cv::String>& decoded_info,
+                     OutputArrayOfArrays straight_qrcode) const override;
+    bool detectAndDecodeMulti(InputArray img, std::vector<cv::String>& decoded_info, OutputArray points,
+                              OutputArrayOfArrays straight_qrcode) const override;
+
+    String decodeCurved(InputArray in, InputArray points, OutputArray straight_qrcode);
+
+    std::string detectAndDecodeCurved(InputArray in, OutputArray points, OutputArray straight_qrcode);
 };
 
-QRCodeDetector::QRCodeDetector() : p(new Impl) {}
+QRCodeDetector::QRCodeDetector() {
+    p = makePtr<ImplContour>();
+}
 
-QRCodeDetector::~QRCodeDetector() {}
+QRCodeDetector& QRCodeDetector::setEpsX(double epsX) {
+    (std::static_pointer_cast<ImplContour>)(p)->epsX = epsX;
```

Comment:
no need braces here:

`std::static_pointer_cast<ImplContour>(p)`

---

checked `dynamic_pointer_cast` is more secure.

Reply:
added `dynamic_pointer_cast` instead `static_pointer_cast`